braindecode/models/spdnet.py

```diff
@@ -0,0 +1,238 @@
+# Authors: Théo Gnassounou <theo.gnassounou@gmail.com>
+# Code inspired from https://github.com/rkobler/TSMNet.git
+#
+# License: BSD (3-clause)
+from geoopt.tensor import ManifoldParameter
+from geoopt.manifolds import Stiefel
+import torch
+import torch.nn as nn
+from .functions import logm, regm
+from .base import EEGModuleMixin
+
+
+class CovLayer(nn.Module):
+    """Covariance layer.
+
+    This class compute the covariance of a batch of
+    symmetric matrices.
+
+    Parameters
+    ----------
+    X : torch.Tensor
+        Batch of symmetric matrices
+
+    Returns
+    -------
+    torch.Tensor
+        Batch of covariance matrices
+    """
+    def forward(self, X):
+        n_batch, n_channels, _ = X.size()
+        torch_covs = torch.empty((n_batch, n_channels, n_channels)).to(X.device)
+        for i, batch in enumerate(X):
+            torch_covs[i] = batch.cov(correction=0)
+        return torch_covs
+
+
+class BiMap(nn.Module):
+    """BiMap layer from [1]_.
+
+    This class generates more compact and discriminative
+    representations of symmetric matrices by projecting them
+    into a lower dimensional subspace Stiefel manifold.
+
+    Parameters
+    ----------
+    input_shape : int
+        Input shape
+    output_shape : int
+        Output shape
+
+    Returns
+    -------
+    torch.Tensor
+        Modified symmetric matrix
+
+    References
+    ----------
+    .. [1] Zhiwu Huang and Luc Van Gool, 2016,
+           A Riemannian Network for SPD Matrix Learning
+           AAAI
+    """
+    def __init__(self, input_shape, output_shape, threshold=1e-4):
+        super(BiMap, self).__init__()
+        self.input_shape_ = input_shape
+        self.output_shape_ = output_shape
+        self.threshold_ = threshold
+        self.manifold_ = Stiefel()
+        assert input_shape >= output_shape
+        self.W = ManifoldParameter(
+            torch.empty([1, input_shape, output_shape]), manifold=self.manifold_
+        )
+        self.reset_parameters()
+
+    @torch.no_grad()
+    def reset_parameters(self):
+        W = torch.rand(self.W.shape, dtype=self.W.dtype, device=self.W.device)
+        s, U = torch.linalg.eigh(W.transpose(-1, -2) @ W)
+        smod = s.clamp(min=self.threshold_).rsqrt()
+        A = U @ torch.diag_embed(smod) @ U.transpose(-1, -2)
+        self.W.data = W @ A
+
+    def forward(self, X):
+        return self.W.transpose(-1, -2) @ X @ self.W
+
+
+class ReEig(nn.Module):
+    """ReEig layer from [1]_.
+
+    This class add non-linearity to the network by
+    applying a rectified linear unit to the eigenvalues
+    of a symmetric matrix.
```

Comment:
Say that by doing this it makes the matrix non-negative and positive if threshold > 0.